### ISSUE TITLE
Rule proposal: `prefer-set-has`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#495]
Enforce use `Set` for existence check over `Array#includes`.

# Fail

```js
const array = ['foo', 'bar'];

if (array.includes('foo')) {}
```

# Pass

```js
const set = new Set(['foo', 'bar']);

if (set.has('foo')) {}
```

This rule might should not fail on this

```js
if (['foo', 'bar'].includes('foo')) {}
```

Maybe only check variables and check references. Also perhaps the rule name `prefer-set-has` ?

real world case: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/d638e5477d7240ea777dcaf249d0121a5a1910eb/rules/prefer-node-append.js#L6-L17

**Comment by yakov116** [comment]
@fisker your would need to know that the array is never used anywhere else. Maybe items are pushed on laster that will have duplicate values.

**Comment by fisker** [comment]
@yakov116 Yes, need check all references only called `#includes()` or `#indexOf()`, and not exported.

**Comment by sindresorhus** [comment]
> Also perhaps the rule name prefer-set-has ?

👍 

> Yes, need check all references only called #includes() or #indexOf(), and not exported.

I don't think it could consider `indexOf` as it's not just used for existence checks, and we already have a rule to enforce `includes` over `indexOf`.

---

I think this would be a useful rule, but I don't think it should be auto-fixable.

**Comment by sindresorhus** [comment]
Accepted

**Comment by fisker** [comment]
@sindresorhus Why do you think it should  not auto-fixable?

I've made a prototype works fine with auto-fix.

**Comment by sindresorhus** [comment]
I guess if you can confidently check all references, it's fine.

**Comment by silverwind** [comment]
What would be the limitations of this rule? Say a external API passes a `Array`. Would the rule require me to convert that to a `Set` first? Or does it only operate on arrays declared via literals?

**Comment by fisker** [comment]
@silverwind I'm going to make it only work on `ArrayExpression` first, so external `Array` will not reported. You can  help review #604

**Comment by silverwind** [comment]
I also wonder for cases where I have a small `Array` which I need to check for existance of an item but which I'd then need to pass on as `Array` to an external consumer. I imagine the conversation to `Set` and back may end up being worse performance than just calling `.includes` on a short array.

**Comment by fisker** [comment]
You can alway use `eslint-disable` to disable it on small array, or we can option to do add length check later.

**Comment by silverwind** [comment]
Yeah of course. I generally try to avoid using `eslint-disable`, but this rule should still be good for manually checking a code base for inefficiencies related to `.includes`.
